fl4health/clients/apfl_client.py

```diff
@@ -1,255 +1,101 @@
-        local_meter = MetricAverageMeter(self.metrics, "local")
-        personal_meter = MetricAverageMeter(self.metrics, "personal")
-        # By default the APFL client trains by epochs
-        metric_values = self.train_by_epochs(local_epochs, global_meter, local_meter, personal_meter)
-        # FitRes should contain local parameters, number of examples on client, and a dictionary holding metrics
-        # calculation results.
-        return (
-            self.get_parameters(config),
-            self.num_examples["train_set"],
-            metric_values,
-        )
-
-    def evaluate(self, parameters: NDArrays, config: Config) -> Tuple[float, int, Dict[str, Scalar]]:
-        if not self.initialized:
-            self.setup_client(config)
-
-        self.set_parameters(parameters, config)
-        # Default APFL uses an average meter
-        global_meter = MetricAverageMeter(self.metrics, "global")
-        local_meter = MetricAverageMeter(self.metrics, "local")
-        personal_meter = MetricAverageMeter(self.metrics, "personal")
-        loss, metric_values = self.validate(global_meter, local_meter, personal_meter)
-        # EvaluateRes should return the loss, number of examples on client, and a dictionary holding metrics
-        # calculation results.
-        return (
-            loss,
-            self.num_examples["validation_set"],
-            metric_values,
-        )
-
-    def _handle_logging(
-        self, loss_dict: Dict[str, float], metrics_dict: Dict[str, Scalar], is_validation: bool = False
-    ) -> None:
-        loss_string = "\t".join([f"{key}: {str(val)}" for key, val in loss_dict.items()])
-        metric_string = "\t".join([f"{key}: {str(val)}" for key, val in metrics_dict.items()])
-        metric_prefix = "Validation" if is_validation else "Training"
-        log(
-            INFO,
-            f"alpha: {self.model.alpha} \n"
-            f"Client {metric_prefix} Losses: {loss_string} \n"
-            f"Client {metric_prefix} Metrics: {metric_string}",
+        super().__init__(
+            data_path, metrics, device, loss_meter_type, metric_meter_type, use_wandb_reporter, checkpointer
         )
+        # Apfl Module which holds both local and global models
+        # and gives the ability to get personal, local and global predictions
+        self.model: APFLModule
 
-    def train_step(self, input: torch.Tensor, target: torch.Tensor) -> ApflTrainStepOutputs:
+        # local_optimizer is used on the local model
+        # Usual self.optimizer is used for global model
+        self.local_optimizer: Optimizer
+
+    def is_start_of_local_training(self, step: int) -> bool:
+        return step == 0
+
+    def update_after_step(self, step: int) -> None:
+        if self.is_start_of_local_training(step) and self.model.adaptive_alpha:
+            self.model.update_alpha()
+
+    def split_optimizer(self, global_optimizer: Optimizer) -> Tuple[Optimizer, Optimizer]:
```

Comment:
I like this idea and it's a great manipulation of the optimizers. However, I wonder if we're overloading the self.optimizer property here a bit by "assuming" or forcing it to be the global optimizer. That is, from an un-aware user, who knows APFL but not our library, it might be surprising to define only a single optimizer where they know that two optimizers **must** be used for APFL. I made a suggestion that we shift from having get_optimizer from returning a single optimizer to a dictionary. We could also consider defining a custom class that may have more than one optimizer in it.

Reply:
Updated in [61693e7](https://github.com/VectorInstitute/FL4Health/pull/62/commits/61693e7a371f5a3b4a0ab9237a87cc0dcf339b2d)